packages/flutter/lib/src/services/text_boundary.dart

```diff
@@ -0,0 +1,224 @@
+// Copyright 2014 The Flutter Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+import 'dart:math' as math;
+import 'dart:ui';
+
+import 'package:characters/characters.dart' show CharacterRange;
+
+import 'text_layout_metrics.dart';
+
+/// An interface for retrieving the logical text boundary (left-closed-right-open)
+/// at a given location in a document.
+///
+/// The input [TextPosition] points a position between 2 code units (which can
+/// be visually represented by the caret if the selection were to collapse to
+/// that position).
+abstract class TextBoundary {
+  /// A constant constructor to be enable subclass override.
+  const TextBoundary();
+
+  /// Returns the leading text boundary at the given location, inclusive.
+  TextPosition getLeadingTextBoundaryAt(TextPosition position);
+
+  /// Returns the trailing text boundary at the given location, exclusive.
+  TextPosition getTrailingTextBoundaryAt(TextPosition position);
+
+  /// Gets the text boundary range that encloses the input position.
+  TextRange getTextBoundaryAt(TextPosition position) {
+    return TextRange(
+      start: getLeadingTextBoundaryAt(position).offset,
+      end: getTrailingTextBoundaryAt(position).offset,
+    );
+  }
+}
+
+/// A text boundary that uses code units as logical boundaries.
+///
+/// This text boundary treats every character in input string as an utf-16 code
+/// unit. This can be useful when handling text without any grapheme cluster,
+/// e.g. the obscure string in [EditableText]. If you are handling text that may
+/// include grapheme clusters, consider using [CharacterBoundary].
+class CodeUnitBoundary extends TextBoundary {
+  /// Creates a [CodeUnitBoundary] with the text.
+  const CodeUnitBoundary(this._text);
+
+
+  final String _text;
+
+  @override
+  TextPosition getLeadingTextBoundaryAt(TextPosition position) {
+    if (position.offset <= 0) {
+      return const TextPosition(offset: 0);
+    }
+    if (position.offset > _text.length ||
+        (position.offset == _text.length && position.affinity == TextAffinity.downstream)) {
+      return TextPosition(offset: _text.length, affinity: TextAffinity.upstream);
+    }
+    switch (position.affinity) {
+      case TextAffinity.upstream:
+        return TextPosition(offset: math.min(position.offset - 1, _text.length));
```

Comment:
Is the `offset -1` strategy correct? If there's no bidi text or soft line wraps in a paragraph, then we have
```
TextPosition(x, downstream) == TextPosition(x, upstream) != TextPosition(x-1, downstream)
```
 right?

Reply:
My understanding of codeUnitboundary is the following. If you have a text 'abc'. To describe the boundary of code unit 'b' will be

leading : TextPosition(offset: 1, affinity:TextAffinity.downStream)
trailing:  TextPosition(offset: 2, affinity:TextAffinity.upStream)

The current code should capture this. It make sure any text position that point to 'b' will return that. WDYT?

Reply:
Yeah according to what's in `TextPainter` that looks correct